### PR TITLE
planner: support keep pruned conds

### PR DESCRIPTION
**Pull request by Reminiscent**
### What problem does this PR solve?

Issue Number: related to https://github.com/pingcap/tidb/issues/33966

Problem Summary:
Add a switch to control whether keep the pruning conditions.

### What is changed and how it works?
Add a switch to control whether keep the pruning conditions.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: sessionctx/variable/tidb_vars.go [hunk 1]
```diff
@@ -740,6 +742,7 @@ const (
 	DefTMPTableSize                       = 16777216
 	DefTiDBEnableLocalTxn                 = false
 	DefTiDBEnableOrderedResultMode        = false
+	DefTiDBKeepPruningConds               = false
```
Comment (Reminiscent) [inline sessionctx/variable/tidb_vars.go:745]:
```suggestion
	DefTiDBKeepPrunedConds               = false
```

File: sessionctx/variable/tidb_vars.go [hunk 2]
```diff
@@ -580,6 +580,8 @@ const (
 
 	// TiDBEnableOrderedResultMode indicates if stabilize query results.
 	TiDBEnableOrderedResultMode = "tidb_enable_ordered_result_mode"
+
+	TiDBKeepPruningConds = "tidb_keep_pruning_conds"
```
Comment (Reminiscent) [inline sessionctx/variable/tidb_vars.go:584]:
```suggestion
	TiDBKeepPrunedConds = "tidb_keep_pruned_conds"
```

File: sessionctx/variable/sysvar.go [hunk 3]
```diff
@@ -1781,6 +1781,10 @@ var defaultSysVars = []*SysVar{
 		s.EnableStableResultMode = TiDBOptOn(val)
 		return nil
 	}},
+	{Scope: ScopeGlobal | ScopeSession, Name: TiDBKeepPruningConds, Value: BoolToOnOff(DefTiDBKeepPruningConds), Hidden: true, Type: TypeBool, SetSession: func(s *SessionVars, val string) error {
```
Comment (Reminiscent) [inline sessionctx/variable/sysvar.go:1784]:
```suggestion
	{Scope: ScopeGlobal | ScopeSession, Name: TiDBKeepPrunedConds, Value: BoolToOnOff(DefTiDBKeepPrunedConds), Hidden: true, Type: TypeBool, SetSession: func(s *SessionVars, val string) error {
```

File: sessionctx/variable/sysvar.go [hunk 4]
```diff
@@ -1781,6 +1781,10 @@ var defaultSysVars = []*SysVar{
...
+		s.KeepPruningConds = TiDBOptOn(val)
```
Comment (Reminiscent) [inline sessionctx/variable/sysvar.go:1785]:
```suggestion
		s.KeepPrunedConds = TiDBOptOn(val)
```

File: sessionctx/variable/session.go [hunk 5]
```diff
@@ -871,6 +871,9 @@ type SessionVars struct {
 	// EnableStableResultMode if stabilize query results.
 	EnableStableResultMode bool
 
+	// KeepPruningConds indicates whether to keep the pruning conditions after partition pruning.
```
Comment (Reminiscent) [inline sessionctx/variable/session.go:874]:
```suggestion
	// KeepPrunedConds indicates whether to keep the pruning conditions after partition pruning.
```

File: sessionctx/variable/session.go [hunk 6]
```diff
@@ -871,6 +871,9 @@ type SessionVars struct {
...
+	KeepPruningConds bool
```
Comment (Reminiscent) [inline sessionctx/variable/session.go:875]:
```suggestion
	KeepPrunedConds bool
```

File: planner/core/rule_partition_processor.go [hunk 7]
```diff
@@ -870,7 +870,7 @@ func (s *partitionProcessor) processRangePartition(ds *DataSource, pi *model.Par
 	if err != nil {
 		return nil, err
 	}
-	if prunedConds != nil {
+	if !ds.ctx.GetSessionVars().KeepPruningConds && prunedConds != nil {
```
Comment (Reminiscent) [inline planner/core/rule_partition_processor.go:873]:
```suggestion
	if !ds.ctx.GetSessionVars().KeepPrunedConds && prunedConds != nil {
```

File: planner/core/integration_test.go [hunk 8]
```diff
@@ -4406,3 +4406,22 @@ func (s *testIntegrationSerialSuite) TestIssue30271(c *C) {
...
+	tk.MustExec("use test")
+	tk.MustExec("set @@session.tidb_partition_prune_mode = 'static'")
+	tk.MustExec("drop table if exists t;")
+	tk.MustExec("create table t(a int, b int, c int, unique index idx(a, b, c)) PARTITION BY RANGE ( `a` ) (PARTITION `PART_202101` VALUES LESS THAN (202102), PARTITION `PART_202102` VALUES LESS THAN (202103), PARTITION `PART_202103` VALUES LESS THAN (202104));")
+	tk.MustExec("set @@tidb_keep_pruning_conds = 1")
+	tk.MustQuery("explain format = 'brief' select count(*) from t where a = 202103 and b = 1 and c = 1;").Check(testkit.Rows(""+
+		"StreamAgg 1.00 root  funcs:count(1)->Column#5",
+		"└─Point_Get 1.00 root table:t, partition:part_202103, index:idx(a, b, c) "))
+
+	tk.MustExec("set @@tidb_keep_pruning_conds = 0")
```
Comment (Reminiscent) [inline planner/core/integration_test.go:4421]:
```suggestion
	tk.MustExec("set @@tidb_keep_pruned_conds = 0")
```

File: planner/core/integration_test.go [hunk 9]
```diff
@@ -4406,3 +4406,22 @@ func (s *testIntegrationSerialSuite) TestIssue30271(c *C) {
...
 	tk.MustQuery("select * from t where (a>'a' and b='a') or (b = 'A' and a < 'd') order by a,c;").Check(testkit.Rows("b a 1", "b A 2", "c a 3"))
 }
+
+func (s *testIntegrationSerialSuite) TestKeepPruningConds(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("set @@session.tidb_partition_prune_mode = 'static'")
+	tk.MustExec("drop table if exists t;")
+	tk.MustExec("create table t(a int, b int, c int, unique index idx(a, b, c)) PARTITION BY RANGE ( `a` ) (PARTITION `PART_202101` VALUES LESS THAN (202102), PARTITION `PART_202102` VALUES LESS THAN (202103), PARTITION `PART_202103` VALUES LESS THAN (202104));")
+	tk.MustExec("set @@tidb_keep_pruning_conds = 1")
```
Comment (Reminiscent) [inline planner/core/integration_test.go:4416]:
```suggestion
	tk.MustExec("set @@tidb_keep_pruned_conds = 1")
```

**Review by chrysan — APPROVED** [review]
LGTM

File: sessionctx/variable/tidb_vars.go [hunk 10]
```diff
@@ -580,6 +580,8 @@ const (
 
 	// TiDBEnableOrderedResultMode indicates if stabilize query results.
 	TiDBEnableOrderedResultMode = "tidb_enable_ordered_result_mode"
+
```
Comment (Reminiscent) [inline sessionctx/variable/tidb_vars.go:583]:
```suggestion

// TiDBKeepPrunedConds indicates whether to keep the pruning conditions after partition pruning.
```

File: executor/set_test.go [hunk 11]
```diff
@@ -520,6 +520,16 @@ func (s *testSerialSuite1) TestSetVar(c *C) {
...
 
+	// test for tidb_keep_pruning_conds
+	tk.MustQuery(`select @@tidb_keep_pruning_conds`).Check(testkit.Rows("0"))
+	tk.MustExec(`set global tidb_keep_pruning_conds = 1`)
+	tk.MustQuery(`select @@global.tidb_keep_pruning_conds`).Check(testkit.Rows("1"))
+	tk.MustExec(`set global tidb_keep_pruning_conds = 0`)
+	tk.MustQuery(`select @@global.tidb_keep_pruning_conds`).Check(testkit.Rows("0"))
+	tk.MustExec(`set tidb_keep_pruning_conds=1`)
+	tk.MustQuery(`select @@global.tidb_keep_pruning_conds`).Check(testkit.Rows("0"))
+	tk.MustQuery(`select @@tidb_keep_pruning_conds`).Check(testkit.Rows("1"))
```
Comment (Reminiscent) [inline executor/set_test.go:531]:
```suggestion
	// test for tidb_keep_pruned_conds
	tk.MustQuery(`select @@tidb_keep_pruned_conds`).Check(testkit.Rows("0"))
	tk.MustExec(`set global tidb_keep_pruned_conds = 1`)
	tk.MustQuery(`select @@global.tidb_keep_pruned_conds`).Check(testkit.Rows("1"))
	tk.MustExec(`set global tidb_keep_pruned_conds = 0`)
	tk.MustQuery(`select @@global.tidb_keep_pruned_conds`).Check(testkit.Rows("0"))
	tk.MustExec(`set tidb_keep_pruning_conds=1`)
	tk.MustQuery(`select @@global.tidb_keep_pruned_conds`).Check(testkit.Rows("0"))
	tk.MustQuery(`select @@tidb_keep_pruned_conds`).Check(testkit.Rows("1"))
```

File: planner/core/integration_test.go [hunk 12]
```diff
@@ -4406,3 +4406,22 @@ func (s *testIntegrationSerialSuite) TestIssue30271(c *C) {
 	tk.MustExec("set names utf8mb4 collate utf8mb4_general_ci;")
 	tk.MustQuery("select * from t where (a>'a' and b='a') or (b = 'A' and a < 'd') order by a,c;").Check(testkit.Rows("b a 1", "b A 2", "c a 3"))
 }
+
+func (s *testIntegrationSerialSuite) TestKeepPruningConds(c *C) {
```
Comment (Reminiscent) [inline planner/core/integration_test.go:4410]:
```suggestion
func (s *testIntegrationSerialSuite) TestKeepPrunedConds(c *C) {
```

File: sessionctx/variable/tidb_vars.go [hunk 13]
```diff
@@ -580,6 +580,8 @@ const (
...
+	TiDBKeepPrunedConds = "tidb_keep_pruned_conds"
```
Comment (Reminiscent) [inline sessionctx/variable/tidb_vars.go:584]:
```suggestion

    // TiDBKeepPrunedConds indicates whether to keep the pruning conditions after partition pruning.
	TiDBKeepPrunedConds = "tidb_keep_pruned_conds"
```

File: executor/set_test.go [hunk 14]
```diff
@@ -520,6 +520,16 @@ func (s *testSerialSuite1) TestSetVar(c *C) {
 	tk.MustQuery(`select @@global.tidb_enable_ordered_result_mode`).Check(testkit.Rows("0"))
 	tk.MustQuery(`select @@tidb_enable_ordered_result_mode`).Check(testkit.Rows("1"))
 
+	// test for tidb_keep_pruned_conds
+	tk.MustQuery(`select @@tidb_keep_pruned_conds`).Check(testkit.Rows("0"))
+	tk.MustExec(`set global tidb_keep_pruned_conds = 1`)
+	tk.MustQuery(`select @@global.tidb_keep_pruned_conds`).Check(testkit.Rows("1"))
+	tk.MustExec(`set global tidb_keep_pruned_conds = 0`)
+	tk.MustQuery(`select @@global.tidb_keep_pruned_conds`).Check(testkit.Rows("0"))
+	tk.MustExec(`set tidb_keep_pruning_conds=1`)
```
Comment (Reminiscent) [inline executor/set_test.go:529]:
```suggestion
	tk.MustExec(`set tidb_keep_pruned_conds=1`)
```

File: sessionctx/variable/session.go [hunk 15]
```diff
@@ -871,6 +871,9 @@ type SessionVars struct {
 	// EnableStableResultMode if stabilize query results.
 	EnableStableResultMode bool
 
+	// KeepPrunedConds indicates whether to keep the pruning conditions after partition pruning.
```
Comment (Reminiscent) [inline sessionctx/variable/session.go:874]:
```suggestion
	// KeepPrunedConds indicates whether to keep the pruned conditions after partition pruning.
```

File: sessionctx/variable/tidb_vars.go [hunk 16]
```diff
@@ -580,6 +580,9 @@ const (
 
 	// TiDBEnableOrderedResultMode indicates if stabilize query results.
 	TiDBEnableOrderedResultMode = "tidb_enable_ordered_result_mode"
+
+    // TiDBKeepPrunedConds indicates whether to keep the pruning conditions after partition pruning.
```
Comment (Reminiscent) [inline sessionctx/variable/tidb_vars.go:584]:
```suggestion
    // TiDBKeepPrunedConds indicates whether to keep the pruned conditions after partition pruning.
```

File: sessionctx/variable/tidb_vars.go [hunk 17]
```diff
@@ -740,6 +743,7 @@ const (
 	DefTMPTableSize                       = 16777216
 	DefTiDBEnableLocalTxn                 = false
 	DefTiDBEnableOrderedResultMode        = false
+	DefTiDBKeepPrunedConds               = false
```
Comment (winoros) [inline sessionctx/variable/tidb_vars.go:746]: Format the space and tab?

**Review by winoros — APPROVED** [review]

**Review by time-and-fate — APPROVED** [review]

**Pull request merged**